### PR TITLE
java versions preview enhancements

### LINKED ISSUE #2340: Confusing attempted safety check with XX_PREVIEW
## What happened?

Using the PRs @iamdanfox wrote, I configured my JDK to use preview versions:

```gradle
javaVersions {
  libraryTarget = '19_PREVIEW'
  distributionTarget = '19_PREVIEW'
  runtime = '19_PREVIEW'
}
```

This failed with error:

```
> Cannot cast object '19_PREVIEW' with class 'java.lang.String' to class 'int'
```

After much gnashing of teeth (including figuring out if I was even using the right version of Baseline) I discovered that while distributionTarget and runtime support receiving strings, libraryTarget does not, and requires an int.

This was extremely confusing to me. I think that if it's desired that libraryTarget not be allowed to publish with preview versions (it makes sense), it should accept the string and give a more pleasant error message. 

## What did you want to happen?

Consistency. Let me write:

```
javaVersions {
  libraryTarget = '19'
  distributionTarget = '19_PREVIEW'
  runtime = '19_PREVIEW'
}
```

and if I do mess up and write '19_PREVIEW' then error with

```
> Preview features cannot be enabled on library projects. This is because it transitively requires all consumers to enable preview features, and given these are subject to change in backwards-breaking ways, this makes published libraries extremely brittle.
> Offending declaration: libraryTarget = '19_PREVIEW'
```

and not a cast exception.

### PR DESCRIPTION
**Pull request by j-baker**
+ support setting preview on a project-by-project basis.
+ Fixes #2340
+ fails more elegantly if javaVersions is set on not-the-root.

### CODE REVIEW

File: gradle-baseline-java/src/main/groovy/com/palantir/baseline/plugins/javaversions/BaselineJavaVersionsExtension.java [hunk 1]
```diff
@@ -61,6 +70,16 @@ public final void setLibraryTarget(int value) {
         libraryTarget.set(JavaLanguageVersion.of(value));
     }
 
+    public final void setLibraryTarget(String value) {
```
Comment (j-baker) [inline gradle-baseline-java/src/main/groovy/com/palantir/baseline/plugins/javaversions/BaselineJavaVersionsExtension.java:67]: this is the bit that solves #2340.

File: gradle-baseline-java/src/test/groovy/com/palantir/baseline/BaselineJavaVersionIntegrationTest.groovy [hunk 2]
```diff
@@ -220,7 +250,7 @@ class BaselineJavaVersionIntegrationTest extends IntegrationSpec {
         when:
         buildFile << '''
         javaVersions {
-            libraryTarget = 11
+            libraryTarget = '11'
```
Comment (j-baker) [inline gradle-baseline-java/src/test/groovy/com/palantir/baseline/BaselineJavaVersionIntegrationTest.groovy:253]: this just tests that the string format for setting versions works.

File: gradle-baseline-java/src/main/groovy/com/palantir/baseline/plugins/javaversions/BaselineJavaVersionsExtension.java [hunk 3]
```diff
@@ -52,6 +54,13 @@ public BaselineJavaVersionsExtension(Project project) {
         runtime.finalizeValueOnRead();
     }
 
+    private static void throwIfNotRootProject(Project project) {
+        if (!project.equals(project.getRootProject())) {
+            throw new GradleException("The javaVersions extension can only be applied to the root project."
+                    + " Did you mean javaVersion, which can be used to override on a project-by-project basis?");
```
Comment (j-baker) [inline gradle-baseline-java/src/main/groovy/com/palantir/baseline/plugins/javaversions/BaselineJavaVersionsExtension.java]: with this I think we'd have probably saved some time!

File: gradle-baseline-java/src/main/groovy/com/palantir/baseline/plugins/javaversions/SubprojectBaselineJavaVersionsExtension.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    public final void setRuntime(int _value) {
+        throw throwCannotSetFromSubproject();
+    }
+
+    @Override
+    public final void setRuntime(String _value) {
+        throw throwCannotSetFromSubproject();
+    }
+
+    private RuntimeException throwCannotSetFromSubproject() {
```
Comment (j-baker) [inline gradle-baseline-java/src/main/groovy/com/palantir/baseline/plugins/javaversions/SubprojectBaselineJavaVersionsExtension.java:61]:
```
FAILURE: Build failed with an exception.

* Where:
Build file '/Volumes/git/thermite/thermite-simulation/build.gradle' line: 2

* What went wrong:
A problem occurred evaluating project ':thermite-simulation'.
> The javaVersions extension can only be used from the root project. Did you mean javaVersion, which can be used to override on a project-by-project basis? You used it from thermite-simulation
```

**Review by CRogers — APPROVED** [review]

**Comment by schlosna** [comment]
This change is causing gradle failures on excavator roll out, e.g. https://github.com/palantir/tritium/pull/1566 with error https://app.circleci.com/pipelines/github/palantir/tritium/2471/workflows/d414b195-beff-486c-9316-e13a206f6981/jobs/18979/parallel-runs/0/steps/0-105

```
FAILURE: Build completed with 2 failures.

1: Task failed with an exception.
-----------
* Where:
Build file '/Volumes/git/palantir/tritium/build.gradle' line: 29

* What went wrong:
A problem occurred evaluating root project 'tritium'.
> failed to access class com.palantir.baseline.plugins.javaversions.BaselineJavaVersionsExtension$LazyJdks from class com.palantir.gradle.jdks.JdksPlugin (com.palantir.baseline.plugins.javaversions.BaselineJavaVersionsExtension$LazyJdks and com.palantir.gradle.jdks.JdksPlugin are in unnamed module of loader org.gradle.internal.classloader.VisitableURLClassLoader @57e989a9)

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
==============================================================================

2: Task failed with an exception.
-----------
* Where:
Build file '/Volumes/git/palantir/tritium/tritium-jmh/build.gradle' line: 17

* What went wrong:
A problem occurred evaluating project ':tritium-jmh'.
> Could not find method api() for arguments [io.dropwizard.metrics:metrics-core] on object of type org.gradle.api.internal.artifacts.dsl.dependencies.DefaultDependencyHandler.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
==============================================================================

* Get more help at https://help.gradle.org

BUILD FAILED in 489ms
On branch roomba/latest-baseline-oss
Your branch is up to date with 'origin/roomba/latest-baseline-oss'.
```

**Comment by schlosna** [comment]
See #2379 for fix

**Pull request merged**